proto/encode.go

```diff
@@ -0,0 +1,68 @@
+package proto
+
+import (
+	"encoding/binary"
+	"io"
+	"unsafe"
+)
+
+// EncodeTag encodes a pair of field number and wire type into a protobuf tag.
+func EncodeTag(f FieldNumber, t WireType) uint64 {
+	return uint64(f)<<3 | uint64(t)
+}
+
+// EncodeZigZag returns v as a zig-zag encoded value.
+func EncodeZigZag(v int64) uint64 {
+	return encodeZigZag64(v)
+}
+
+func encodeZigZag64(v int64) uint64 {
+	return uint64((v << 1) ^ (v >> 63))
+}
+
+func encodeZigZag32(v int32) uint32 {
+	return uint32((v << 1) ^ (v >> 31))
+}
+
+type encodeFunc = func([]byte, unsafe.Pointer, flags) (int, error)
+
+func encodeVarint(b []byte, v uint64) (int, error) {
+	i := 0
+	for v >= 0x80 {
```

Comment:
It might be worth unrolling this loop, e.g. have a look at [this](https://github.com/gogo/protobuf/blob/09ab7735f7757c093f5b0a2285bff3998d684a61/proto/table_marshal.go#L1521).

You should be able to use something like `(math_bits.Len64(v|1) + 6) / 7` to determine how many bytes you'll need in advance (so you can return `io.ErrShortBuffer` if the buffer doesn't have enough space).